teps/0138-improve-tekton-pipeline-feature-flags.md

```diff
@@ -0,0 +1,73 @@
+
+- When promoting features, it could cause confusions for contributors to be dependent on the fact whether an apiVersion is available. For example, during [the promotion to beta for projected workspaces](https://github.com/tektoncd/pipeline/pull/5530), the v1 api's existence led to confusions of what to do with beta features in v1beta1 and its difference with in v1.
+
+## Goals
+
+- [Decouple API versioning and feature stability](https://github.com/tektoncd/pipeline/issues/6592)
+  - Feature validation and implementation should be the same for all API versions.
+- Cluster operators can enable an individual feature without being forced to opt into other features at the same stability level.
+  - Increase the visibility of stability levels for features that have their own feature flags and ones that are not tied with the api spec
+- Describe a testing strategy that will give us confidence in our implementation
+  - One of the main reasons why [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-feature-gates.md#pros-and-cons) chose the existing style of feature flags was to have a simpler testing matrix.  With upgrades to the feature versioning, we need to ensure that our testing strategy gives us confidence in our feature coverage.
+  - Restructure conditions for integration tests to avoid  [Skipped Integration Tests](https://github.com/tektoncd/pipeline/issues/6079) based on different required sets of behavioral feature flags in certain stability levels
+- Minimize backwards incompatible changes and provide simple migration paths for cluster operators.
+- Cluster operators can set a default feature stability level.
+  - For example, we want to address the current case where a cluster operator only wants to use `stable` features and we have defaulted `beta` as the opt-in features stability then they would have to change back the default feature stability to use all `stable` features by reading through the release notes.
+- Build consistency in the usage for the group feature flag and the behavioral feature flags.
+
+## Non goals
+
+- Better guidance on feature promotion and when features can be promoted
+  - This is a nice-to-have but not necessarily a blocker, since the feature graduating process should not affect the implementation of how features are enabled.
+- Separation of controls for cluster operators and end users
+  - On a cluster that serves multiple teams, if a cluster operator enables/disables one feature then all the teams are affected. This might break pipelines/tasks for certain teams that rely on a feature. For example, if the cluster operator transitions from `alpha` to `beta` or `stable` features, then the users that are currently relying on `alpha` features can no longer use them. 
```

Comment:
If our users ask for more flexibility, perhaps we could consider looking into [OpenFeature](https://openfeature.dev/docs/reference/intro).

Reply:
I looked into [OpenFeature](https://openfeature.dev/docs/reference/intro). However, it was to see if it could help with our testing matrix if we wanted to switch to per-feature flags. However, since we descoped per-feature flags from this proposal, it might make sense to take this conversation there when we open a TEP on per-feature flags. 
To close the loop here, I think [TEP 0085](https://github.com/tektoncd/community/blob/main/teps/0085-per-namespace-controller-configuration.md) addresses this issue better.